collectors/modality-probe-collector-common/src/lib.rs

```
@@ -283,320 +354,414 @@ impl TryFrom<&LogFileRow> for LogEntry {
     }
-
-    next_segment_id
 }
 
-pub fn add_owned_report_to_entries(
-    report: LogReport,
-    session_id: SessionId,
-    initial_segment_id: SegmentId,
-    receive_time: DateTime<Utc>,
-    log_entries_buffer: &mut Vec<LogEntry>,
-) -> u32 {
-    let mut next_segment_id = initial_segment_id.0;
-    let probe_id = report.probe_id;
-    for segment in report.segments {
-        let mut segment_index = 0;
-
-        for clock_bucket in segment.clocks {
-            log_entries_buffer.push(LogEntry {
-                session_id,
-                segment_id: next_segment_id.into(),
-                segment_index,
-                probe_id,
-                data: LogEntryData::LogicalClock(
-                    clock_bucket.id,
-                    clock_bucket.epoch,
-                    clock_bucket.ticks,
-                ),
-                receive_time,
-            });
+enum Next {
+    Clock(ProbeId),
+    Payload(EventId),
+    DontKnow,
+}
 
-            segment_index += 1;
+impl Report {
+    pub fn write_into_le_bytes(&self, bytes: &mut [u8]) -> Result<usize, ReportWireError> {
+        let mut wire = WireReport::new_unchecked(bytes);
+        wire.check_len()?;
+        wire.set_fingerprint();
+        wire.set_probe_id(self.probe_id);
+        wire.set_clock(modality_probe::pack_clock_word(
+            self.probe_clock.epoch,
+            self.probe_clock.ticks,
+        ));
+        wire.set_seq_num(self.seq_num);
+        wire.set_n_clocks(self.frontier_clocks.len() as _);
+
+        let num_u32_entries: usize = self
+            .event_log
+            .iter()
+            .map(|e| match e {
+                EventLogEntry::Event(_) => 1,
+                EventLogEntry::EventWithPayload(_, _) => 2,
+                EventLogEntry::TraceClock(_) => {
+                    mem::size_of::<LogicalClock>() / mem::size_of::<u32>()
+                }
+            })
+            .sum();
+
+        wire.set_n_log_entries(num_u32_entries as _);
```

Same concern re: `as` casting causing panics rather than returning errors.